packages/vaadin-grid/src/arrayDataProvider.js

```diff
@@ -0,0 +1,142 @@
+ */
+function checkPaths(arrayToCheck, action, items) {
+  if (!items.length) {
+    return false;
+  }
+
+  let result = true;
+
+  for (let i in arrayToCheck) {
+    const path = arrayToCheck[i].path;
+
+    // skip simple paths
+    if (!path || path.indexOf('.') === -1) {
+      continue;
+    }
+
+    const parentProperty = path.replace(/\.[^.]*$/, ''); // a.b.c -> a.b
+    if (get(parentProperty, items[0]) === undefined) {
+      console.warn(`Path "${path}" used for ${action} does not exist in all of the items, ${action} is disabled.`);
+      result = false;
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Sorts the given array of items based on the sorting rules and returns the result.
+ *
+ * @param {Array<any>} items
+ * @param {Array<GridSorter>} items
+ * @return {Array<any>}
+ */
+function multiSort(items, sortOrders) {
+  return items.sort((a, b) => {
+    return sortOrders
+      .map((sortOrder) => {
+        if (sortOrder.direction === 'asc') {
+          return compare(get(sortOrder.path, a), get(sortOrder.path, b));
+        } else if (sortOrder.direction === 'desc') {
+          return compare(get(sortOrder.path, b), get(sortOrder.path, a));
+        }
+        return 0;
+      })
+      .reduce((p, n) => {
+        return p ? p : n;
+      }, 0);
+  });
+}
+
+/**
+ * @param {unknown} value
+ * @return {string}
+ */
+function normalizeEmptyValue(value) {
+  if ([undefined, null].indexOf(value) >= 0) {
+    return '';
+  } else if (isNaN(value)) {
+    return value.toString();
+  } else {
+    return value;
+  }
+}
+
+/**
+ * @param {unknown} a
+ * @param {unknown} b
+ * @return {number}
+ */
+function compare(a, b) {
+  a = normalizeEmptyValue(a);
+  b = normalizeEmptyValue(b);
+
+  if (a < b) {
+    return -1;
+  }
+  if (a > b) {
+    return 1;
+  }
+  return 0;
+}
+
+/**
+ * @param {!Array<!GridItem>} items
+ * @return {!Array<!GridItem>}
+ */
+function filter(items, filters) {
+  return items.filter((item) => {
+    return (
+      filters.filter((filter) => {
+        const value = normalizeEmptyValue(get(filter.path, item));
+        const filterValueLowercase = normalizeEmptyValue(filter.value).toString().toLowerCase();
+        return value.toString().toLowerCase().indexOf(filterValueLowercase) === -1;
+      }).length === 0
+    );
+  });
+}
+
+/**
+ * WARNING: This API is still intended for internal purposes only and
+ * may change any time.
+ *
+ * Creates a new grid compatible data provider that serves the items
+ * from the given array as data when requested by the grid.
+ *
+ * @param {Array<any>} items
+ * @return {GridDataProvider<any>}
+ */
+export const createArrayDataProvider = (allItems) => {
+  return (params, callback) => {
+    let items = allItems ? [...allItems] : [];
+    if (params.filters && checkPaths(params.filters, 'filtering', items)) {
+      items = filter(items, params.filters);
+    }
+
+    if (params.sortOrders.length && checkPaths(params.sortOrders, 'sorting', items)) {
```

Comment:
It is guaranteed by data provider typing https://github.com/vaadin/web-components/blob/master/packages/vaadin-grid/src/interfaces.d.ts#L23 Added the check anyhow since we still use js internally